builtin/credential/aws/client.go

```diff
@@ -80,20 +81,39 @@ func (b *backend) getClientConfig(s logical.Storage, region, stsRole, clientType
 		return nil, fmt.Errorf("could not compile valid credentials through the default provider chain")
 	}
 
+	stsConfig, err := b.getRawClientConfig(s, region, "sts")
```

Comment:
I'm not sure how to answer your question because the phrase "STS creds" isn't well defined, so I'm not sure what you're asking and thus not sure how to answer. Let me try to explain a little more, and if I'm still not answering your question, it'd probably be best to discuss offline. Apologies if this response is a bit basic, but I'm just trying to understand where we're not connecting.

STS is both a service that requires credentials to authenticate to it and a service that can act as a source of credentials. The code in question does both. I'm pretty sure that any type of IAM entity can authenticate to STS.

> it's safe to assume that you always want to use STS creds to make the client calls

Which client calls?

If an `stsRole` has been configured for a given target account ID, then we use the existing creds Vault has been configured with to authenticate to STS to ask for a new set of credentials corresponding to the `stsRole`, and it uses those credentials for future requests. I'm not really changing this code path at all -- I'm just pulling the existing code up a level so the results can be used, but no material changes are being made. And we're depending on customers to configure the IAM roles such that Vault is able to retrieve credentials via `sts:AssumeRole`, so the customer has said he/she wants to get credentials from STS.

If no `stsRole` has been configured for an account (the big change I'm making here), then we just use the existing creds Vault has been configured with to authenticate to `sts:GetCallerIdentity` to learn the identity of the creds being given; we're not asking for any new credentials from STS. And we continue to use the creds Vault has been given (i.e., not any temporary credentials retrieved via STS) to make future client calls.

> it's safe to assume that you always want to use STS creds to make the client calls

Which client calls? The only times we ask STS for credentials is when clients have explicitly told Vault to go to STS for credentials for those calls, and I'm not changing that.

> I assume that in any given account we should always be able to get STS creds? 

In any given account, we should always be able to call `sts:GetCallerIdentity` using any valid creds, whether those are IAM user creds or assumed role creds. Beyond that, it depends. If users have set up the appropriate STS configs in Vault and have set up the appropriate IAM permissions in AWS, then yes. But if's really a user configuration issue.

> Or will this not be the case if they're not using IAM roles

The `stsRole` must be an IAM role, yes. Otherwise, even if Vault is using, say, IAM User credentials, all this will still work. IAM users can call `sts:GetCallerIdentity` and IAM users can call `sts:AssumeRole` (as long as the appropriate IAM permissions have been set up).

Does this answer what you're asking? If not, what am I missing?